src/group_theory/monoid_localization.lean

```diff
@@ -266,25 +306,442 @@ show _ * _ = _, by rw [mul_inv_left, mul_one]
+  k.1 (g (sec S f.1 z).2) * f.map hy k z = k.1 (g (sec S f.1 z).1) :=
+by rw [mul_comm, f.map_mul_right]
+
+@[simp, to_additive] lemma map_id (z : N) :
+  f.map (λ y, show monoid_hom.id M y ∈ S, from y.2) f z = z :=
+f.lift_id z
+
+/-- If `comm_monoid` homs `g : M →* P, l : P →* A` induce maps of localizations, the composition
+    of the induced maps equals the map of localizations induced by `l ∘ g`. -/
+@[to_additive "If `add_comm_monoid` homs `g : M →+ P, l : P →+ A` induce maps of localizations, the composition of the induced maps equals the map of localizations induced by `l ∘ g`."]
+lemma map_comp_map {A : Type*} [comm_monoid A] {U : submonoid A} {R} [comm_monoid R]
+  (j : localization_map U R) {l : P →* A} (hl : ∀ w : T, l w ∈ U) :
+  (k.map hl j).comp (f.map hy k) = f.map (λ x, show l.comp g x ∈ U, from hl ⟨g x, hy x⟩) j :=
+begin
+  ext z,
+  show j.1 _ * _ = j.1 (l _) * _,
+  { rw [mul_inv_left, ←mul_assoc, mul_inv_right],
+    show j.1 _ * j.1 (l (g _)) = j.1 (l _) * _,
+    rw [←j.1.map_mul, ←j.1.map_mul, ←l.map_mul, ←l.map_mul],
+    exact k.comp_eq_of_eq hl j
+      (by rw [k.1.map_mul, k.1.map_mul, sec_spec' k.3, mul_assoc, map_mul_right]) },
+end
+
+/-- If `comm_monoid` homs `g : M →* P, l : P →* A` induce maps of localizations, the composition
+    of the induced maps equals the map of localizations induced by `l ∘ g`. -/
+@[to_additive "If `add_comm_monoid` homs `g : M →+ P, l : P →+ A` induce maps of localizations, the composition of the induced maps equals the map of localizations induced by `l ∘ g`."]
+lemma map_map {A : Type*} [comm_monoid A] {U : submonoid A} {R} [comm_monoid R]
+  (j : localization_map U R) {l : P →* A} (hl : ∀ w : T, l w ∈ U) (x) :
+  k.map hl j (f.map hy k x) = f.map (λ x, show l.comp g x ∈ U, from hl ⟨g x, hy x⟩) j x :=
+by rw ←f.map_comp_map hy j hl; refl
+
+variables {g}
+
+/-- If `f : M →* N` and `k : M →* R` are localization maps for a submonoid `S`, we get an
+    isomorphism of `N` and `R`. -/
+@[to_additive "If `f : M →+ N` and `k : M →+ R` are localization maps for a submonoid `S`, we get an isomorphism of `N` and `R`."]
+noncomputable def mul_equiv_of_localizations
+  (k : localization_map S P) : N ≃* P :=
+⟨f.lift k.2, k.lift f.2, f.lift_left_inverse, k.lift_left_inverse, monoid_hom.map_mul _⟩
+
+@[to_additive, simp] lemma mul_equiv_of_localizations_apply
+  {k : localization_map S P} {x} : f.mul_equiv_of_localizations k x = f.lift k.2 x := rfl
+
+@[to_additive, simp] lemma mul_equiv_of_localizations_symm_apply
+  {k : localization_map S P} {x} : (f.mul_equiv_of_localizations k).symm x = k.lift f.2 x := rfl
+
+/-- If `f : M →* N` is a localization map for a submonoid `S` and `k : N ≃* P` is an isomorphism
+    of `comm_monoid`s, `k ∘ f` is a localization map for `M` at `S`. -/
+@[to_additive "If `f : M →+ N` is a localization map for a submonoid `S` and `k : N ≃+ P` is an isomorphism of `add_comm_monoid`s, `k ∘ f` is a localization map for `M` at `S`."]
+def to_mul_equiv (k : N ≃* P) :
```

Comment:
`to_mul_equiv` sounds like it should make a `mul_equiv`

Also, @jcommelin do you think this might cause problems when in practice there is some canonical map M->P floating around that is eq but not defeq to the map given by this definition.

Reply:
That will probably be painful, yes... But maybe we should save that for a later refactor, because this is something that we don't have a good solution for anyways...